packages/repository/src/define-repository-class.ts

```diff
@@ -0,0 +1,192 @@
+  R extends CrudRepository<M>
+> {
+  /**
+   * The constructor for the generated repository class
+   * @param dataSource - DataSource object
+   */
+  new (dataSource: juggler.DataSource): R;
+  prototype: R;
+}
+
+/**
+ * Signature for EntityCrudRepository classes
+ *
+ * @typeParam E - An entity class
+ * @typeParam IdType - ID type for the entity
+ * @typeParam Relations - Relations for the entity
+ */
+export interface EntityCrudRepositoryClass<
+  E extends Entity,
+  IdType,
+  Relations extends object
+> extends CrudRepositoryClass<E, EntityCrudRepository<E, IdType, Relations>> {}
+
+/**
+ * Signature for repository classes that can be used as the base class for
+ * `define*` functions
+ *
+ * @typeParam M - Model class
+ * @typeParam R - Repository class/interface
+ */
+export interface BaseRepositoryClass<M extends Model, R extends Repository<M>> {
+  /**
+   * The constructor for the generated repository class
+   * @param dataSource - DataSource object
+   */
+  new (modeCtor: Constructor<M>, dataSource: juggler.DataSource): R;
+  prototype: R;
+}
+/**
+ * Create (define) a CRUD repository class for the given model.
+ *
+ * @example
+ *
+ * ```ts
+ * const AddressRepository = defineCrudRepositoryClass(Address);
+ * ```
+ *
+ * @param modelClass - A model class such as `Address`.
+ *
+ * @typeParam M - Model class
+ * @typeParam R - CRUD Repository class/interface
+ */
+export function defineCrudRepositoryClass<
+  M extends Model,
+  R extends CrudRepository<M>
+>(
+  modelClass: typeof Model & {prototype: M},
+  baseRepositoryClass: BaseRepositoryClass<M, R>,
+): CrudRepositoryClass<M, R> {
+  const repoName = modelClass.name + 'Repository';
+  const defineNamedRepo = new Function(
+    'ModelCtor',
+    'BaseRepository',
+    `return class ${repoName} extends BaseRepository {
+      constructor(dataSource) {
+        super(ModelCtor, dataSource);
+      }
+    };`,
+  );
+
+  const repo = defineNamedRepo(modelClass, baseRepositoryClass);
+  assert.equal(repo.name, repoName);
+  return repo;
+}
+
+/**
+ * Create (define) an entity CRUD repository class for the given model.
+ *
+ * @example
+ *
+ * ```ts
+ * const ProductRepository = defineEntityCrudRepositoryClass(Product);
+ * ```
+ *
+ * @param entityClass - An entity class such as `Product`.
+ * @param baseRepositoryClass - Base repository class. Defaults to `DefaultCrudRepository`
+ *
+ * @typeParam E - An entity class
+ * @typeParam IdType - ID type for the entity
+ * @typeParam Relations - Relations for the entity
+ */
+export function defineEntityCrudRepositoryClass<
+  E extends Entity,
+  IdType,
+  Relations extends object = {}
+>(
+  entityClass: typeof Entity & {prototype: E},
+  baseRepositoryClass: BaseRepositoryClass<
+    E,
+    EntityCrudRepository<E, IdType, Relations>
+  > = (DefaultCrudRepository as unknown) as BaseRepositoryClass<
+    E,
+    EntityCrudRepository<E, IdType, Relations>
+  >,
+): EntityCrudRepositoryClass<E, IdType, Relations> {
```

Comment:
I think this is not going to support additional methods that could be provided by `baseRepositoryClass` on top of `EntityCrudRepository` interface (e.g. `findByTitle`). A possible solution is to introduce another generic argument `R`, similarly to what `defineCrudRepositoryClass` does.

Please make sure to add a test to verify this use case.